cpp/src/arrow/CMakeLists.txt

```diff
@@ -805,6 +806,10 @@ if(ARROW_IPC)
   add_arrow_test(extension_type_test)
 endif()
 
+if(ARROW_JSON)
+  add_arrow_test(canonical_extension_test SOURCES extension/fixed_shape_tensor_test.cc)
```

Comment:
I think it shouldn't be (that one was added for the install of headers, this one is for adding the tests, if I am not mistaken)

Reply:
It seems that ARM64 Ubuntu 20.04 C++ is running `fixed_shape_tensor_test.cc` but can't find the extension folder.
[See](https://github.com/apache/arrow/actions/runs/4566566697/jobs/8059249761?pr=8510#step:7:2916).

Reply:
I reintroduced `ARROW_WITH_JSON` which seems to have resolved [missing library issue](https://github.com/apache/arrow/actions/runs/4566566697/jobs/8059249761?pr=8510#step:7:2916) that appears when using `ARROW_JSON` instead.

Reply:
Does it work as well if you do `add_definitions(-DARROW_JSON)`?  (to keep the names consistent)

Reply:
I'll test. I wonder why just using `ARROW_JSON` won't work. Is something manipulating `ARROW_JSON` flag at compile time?

Reply:
`add_definitions(-DARROW_JSON)` won't do it as it redefines a flag. See [here](https://github.com/apache/arrow/actions/runs/4574717127/jobs/8076673684?pr=8510#step:11:1247).
I'll revert to `ARROW_WITH_JSON` but I'm still open to other options. Also not we use `ARROW_WITH_X` in several places so this would not be that inconsistent.

Reply:
Strange thing is that we are already using `#ifdef ARROW_JSON` elsewhere:

https://github.com/apache/arrow/blob/8efb43ddcb908747c64cc5048930209305bcdc73/cpp/src/arrow/public_api_test.cc#L50

> `add_definitions(-DARROW_JSON)` won't do it as it redefines a flag.

Ah, because we already have it in `config.h.cmake`. But then indeed it should already work?

Reply:
You might need to include `#include "arrow/config.h"` to be able to use the config ifdef

Reply:
Ha, that solves it locally, thanks!